pvlib/bifacial/utils.py

```diff
@@ -0,0 +1,149 @@
+    projected to the plane defined by the zenith vector and surface_azimuth.
+
+    .. math::
+        \\tan \\phi = \\cos\\left(\\text{solar azimuth}-\\text{system azimuth}
+        \\right)\\tan\\left(\\text{solar zenith}\\right)
+
+    Parameters
+    ----------
+    solar_zenith : numeric
+        Solar zenith angle. [degree].
+    solar_azimuth : numeric
+        Solar azimuth. [degree].
+    surface_azimuth : numeric
+        Azimuth of the module surface, i.e., North=0, East=90, South=180,
+        West=270. [degree]
+
+    Returns
+    -------
+    tan_phi : numeric
+        Tangent of the angle between vertical and the projection of the
+        sun direction onto the YZ plane.
+    """
+    rotation = solar_azimuth - surface_azimuth
+    tan_phi = cosd(rotation) * tand(solar_zenith)
+    return tan_phi
+
+
+def _unshaded_ground_fraction(surface_tilt, surface_azimuth, solar_zenith,
+                              solar_azimuth, gcr, max_zenith=87):
+    r"""
+    Calculate the fraction of the ground with incident direct irradiance.
+
+    .. math::
+        F_{gnd,sky} = 1 - \min{\left(1, \text{GCR} \left|\cos \beta +
+        \sin \beta \tan \phi \right|\right)}
+
+    where :math:`\beta` is the surface tilt and :math:`\phi` is the angle
+    from vertical of the sun vector projected to a vertical plane that
+    contains the row azimuth `surface_azimuth`.
+
+    Parameters
+    ----------
+    surface_tilt : numeric
+        Surface tilt angle. The tilt angle is defined as
+        degrees from horizontal, e.g., surface facing up = 0, surface facing
+        horizon = 90. [degree]
+    surface_azimuth : numeric
+        Azimuth of the module surface, i.e., North=0, East=90, South=180,
+        West=270. [degree]
+    solar_zenith : numeric
+        Solar zenith angle. [degree].
+    solar_azimuth : numeric
+        Solar azimuth. [degree].
+    gcr : float
+        Ground coverage ratio, which is the ratio of row slant length to row
+        spacing (pitch). [unitless]
+    max_zenith : numeric, default 87
+        Maximum zenith angle. For solar_zenith > max_zenith, unshaded ground
+        fraction is set to 0. [degree]
+
+    Returns
+    -------
+    f_gnd_beam : numeric
+        Fraction of distance betwen rows (pitch) with direct irradiance
+        (unshaded). [unitless]
+
+    References
+    ----------
+    .. [1] Mikofski, M., Darawali, R., Hamer, M., Neubert, A., and Newmiller,
+       J. "Bifacial Performance Modeling in Large Arrays". 2019 IEEE 46th
+       Photovoltaic Specialists Conference (PVSC), 2019, pp. 1282-1287.
+       doi: 10.1109/PVSC40753.2019.8980572.
+    """
+    tan_phi = _solar_projection_tangent(solar_zenith, solar_azimuth,
+                                        surface_azimuth)
+    f_gnd_beam = 1.0 - np.minimum(
+        1.0, gcr * np.abs(cosd(surface_tilt) + sind(surface_tilt) * tan_phi))
+    np.where(solar_zenith > max_zenith, 0., f_gnd_beam)  # [1], Eq. 4
```

Comment:
@mikofski I *think* this line has no effect as there is no return value.

Reply:
That's a point, well spotted @kdebrab ! Would you like to open a PR fixing it?

Reply:
This line is probably supposed to start with `f_gnd_beam =` and clip it at max zenith so values don’t blow up.
ICYMI @cwhanse  @kandersolar

Reply:
I agree with @mikofski, that was the intent.

Reply:
these [lines](https://github.com/mikofski/pvlib-python/blob/0428fbef5db15687b746cdc1aa5732b943f0f17d/pvlib/bifacial/utils.py#L88C3-L91C54) are still unchanged in main:
```python
    f_gnd_beam = 1.0 - np.minimum(
        1.0, gcr * np.abs(cosd(surface_tilt) + sind(surface_tilt) * tan_phi))
    np.where(solar_zenith > max_zenith, 0., f_gnd_beam)  # [1], Eq. 4
    return f_gnd_beam  # 1 - min(1, abs()) < 1 always
```
was a PR forthcoming?

Reply:
@mikofski has not been fixed, should have its own issue.